inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+            return 'Sparse With Sharing'
+        elif self.value == FeatureFlag.SPARSE_SHARED_BASELINE_FP16:
+            return 'Sparse Without Sharing'
+        elif self.value == FeatureFlag.SPARSE_FUSED_FP16:
+            return 'Sparse Fused FP16'
+        elif self.value == FeatureFlag.SPARSE_FUSED_FP32:
+            return 'Sparse Fused FP32'
+        elif self.value == FeatureFlag.SPARSE_FUSED_FP32_EXPERIMENTAL:
+            return 'Sparse Fused FP32 Experimental'
+        else:
+            raise 'Pretty print not found for value {self.value}'
+
+
+def spqr_mul(spqr_device: SPQRModule, x, y, feature_flag: FeatureFlag):
+    spqr_cuda.spqr_mul(
+        spqr_device.m,
+        spqr_device.n,
+        spqr_device.bits,
+        spqr_device.beta1,
+        spqr_device.beta2,
+        spqr_device.buff0,
+        spqr_device.row_offsets,
+        spqr_device.col_vals,
+        spqr_device.nnz,
+        x,
+        y,
+        int(feature_flag)
+    )
+
+
+def torch_mul_timer(deq_w, x, num_runs):
+    if len(deq_w.shape) == 1:
+        n = x.shape[0]
+        m = deq_w.shape[0] // n
+    else:
+        m, n = deq_w.shape
+
+    assert (n == x.shape[0])
+
+    runs = torch.empty(num_runs).cpu().float()
+
+    y = torch.zeros(m, dtype=x.dtype, device=x.device)
+
+    for i in range(num_runs):
+        y = torch.zeros_like(y)
+        spqr_cuda.torch_mul_fp16(m, n, deq_w, x, y, runs[i])
+
+    return y, runs
+
+
+def spqr_mul_timer(spqr_device: SPQRModule, x, feature_flag: FeatureFlag, num_runs):
+    runs = torch.empty(num_runs).cpu().float()
+    y = torch.zeros(spqr_device.m, dtype=x.dtype, device=x.device)
+
+    for i in range(num_runs):
+        y = torch.zeros_like(y)
+        spqr_cuda.spqr_mul_timer(
+            spqr_device.m,
+            spqr_device.n,
+            spqr_device.bits,
+            spqr_device.beta1,
+            spqr_device.beta2,
+            spqr_device.buff0,
+            spqr_device.row_offsets,
+            spqr_device.col_vals,
+            spqr_device.nnz,
+            x,
+            y,
+            runs[i],
+            feature_flag)
+
+    return y, runs
+
+
+def write_tensor(spqr_module: SPQRModule, path: str):
+    torch.save(spqr_module, path)
+
+
+def load_compressed_tensor(p: str) -> SPQRModule:
+    spqr_module = torch.load(p)
+    return spqr_module
+
+
+def calculate_buffer_sizes(m, n, beta1, beta2):
+    total_blocks = updiv(m, beta1) * updiv(n, beta2)
+    per_tile_row = 1
+    block_size = beta1 * per_tile_row
+    return block_size * total_blocks
+
+
+def allocate_compressed_buffers(m, n, beta1, beta2, device) -> torch.Tensor:
+    sz0 = calculate_buffer_sizes(m, n, beta1, beta2)
+    return torch.zeros(sz0, dtype=torch.int64, device=device)
+
+
+def num_tiles(m, n, beta1, beta2):
+    return (m // beta1) * (n // beta2)
+
+
+def num_first_order_grups(m, n, beta1, beta2):
+    return num_tiles(m, n, beta1, beta2) * beta1
+
+
+def generate_random_nbit_tensor(sz, bits=3):
```

Comment:
what this function for?

Reply:
Removed - used to be part of the internal test API.